bin/cheribsdtest/cheribsdtest_cheriabi.c

```diff
@@ -428,3 +429,65 @@ CHERIBSDTEST(cheriabi_shmdt_invalid_ptr,
 	free_adjacent_mappings_shm(&mappings);
 	cheribsdtest_success();
 }
+
+CHERIBSDTEST(cheriabi_pipe_oob_write,
+    "Check that writes to a pipe obey bounds")
+{
+	char buf[8192];
+	int fds[2], res;
+	pid_t pid;
+
+	CHERIBSDTEST_CHECK_SYSCALL(pipe(fds));
+	memset(buf, '.', sizeof(buf));
+
+	CHERIBSDTEST_CHECK_SYSCALL(pid = fork());
+	if (pid != 0) {
+		close(fds[1]);
+		CHERIBSDTEST_CHECK_SYSCALL(write(fds[0], buf, sizeof(buf)));
+		CHERIBSDTEST_CHECK_CALL_ERROR(
+		    write(fds[0], buf, sizeof(buf) + 1), EFAULT);
+		close(fds[0]);
+
+		waitpid(pid, &res, 0);
+		CHERIBSDTEST_CHECK_EQ_BOOL(WIFEXITED(res), true);
+		cheribsdtest_success();
+	} else {
+		close(fds[0]);
+		read(fds[1], buf, sizeof(buf));
+		read(fds[1], buf, sizeof(buf));
+		read(fds[1], buf, 1);
```

Comment:
Presumably I'm missing something obvious, but how does this work? The parent's only successfully written `sizeof(buf)` bytes.

Reply:
The 2nd and 3rd read calls return 0 because the descriptor has been closed. I wanted the reads there to absorb the write if the kernel didn't correctly check the capability bounds and thus the writes succeed. I'll add comments to that effect.